### ISSUE TITLE
fscrypt prevents ssh public key authentication even if home directory has been unlocked

### ISSUE THREAD

**Issue by RomanValov** [google/fscrypt#116]
I've set up fscrypt home directory encryption on Ubuntu 18.04 following guide by @tlbdk.

Then I've tried to access my Ubuntu system using ssh public key authentication. Logically ssh server refuses to authorize using public key and forces ssh client to ask user for a password. However in contrast to pam_ecryptfs/pam_mount solutions password is still required for following simultaneous login attempts when home directory has been unlocked.

Having to say if ssh server is restarted after home directory has been unlocked -- public key authentication works.

**Comment by obadz** [comment]
That's annoying.

Also it'd be nice if there was a way to have the "clear version" of the directory (which would just contain your `~/.ssh/authorized_keys` as well as the encrypted version which contains everything so that you can still log into your account even if home not mounted yet (ecryptfs does do that).

**Comment by Redsandro** [comment]
@RomanValov commented:
> Having to say if ssh server is restarted after home directory has been unlocked -- public key authentication works.

This is actually not true; it will only accidentally read the files if they are still in the file system cache. I have experienced the same issue, including experiencing the ssh connection works as long as the user is already logged in. However, 24 hours later, I can no longer login. Running sshd in debug mode reveals why:

```
Aug 29 18:50:04 red sshd[20743]: debug1: trying public key file /home/redsandro/.ssh/authorized_keys
Aug 29 18:50:04 red sshd[20743]: debug1: Could not open authorized keys '/home/redsandro/.ssh/authorized_keys': Required key not available
```

__Could not open authorized keys '/home/sander/.ssh/authorized_keys': Required key not available__

This is the same issue as #128 and I assume the [kernel patches](https://git.kernel.org/pub/scm/fs/fscrypt/fscrypt.git/log/) by @ebiggers will address this. But perhaps @ebiggers can comment on that.

Old `ecryptfs` logged in encrypted home ssh works because the keys are available permanently, not per user per read request (or something). The trick was just to stay logged in (e.g. a `screen` session).

For `fscrypt` until upstream patches are accepted, we need a [workaround](https://stephenreescarter.net/encrypted-home-directories-ssh-key-authentication/).

**Comment by ebiggers** [comment]
Yes, this (along with other issues where system services and containers can't access encrypted files) will be fixed by putting the keys in a filesystem-level keyring instead of a user keyring, the [kernel side of which](https://git.kernel.org/pub/scm/fs/fscrypt/fscrypt.git/log/) is planned to be merged for Linux v5.4.  After that, there will still be some significant changes to the `fscrypt` tool needed, which I'm [working on](https://github.com/ebiggers/fscrypt/commits/fscrypt-key-mgmt-improvements).  For existing encrypted directories I'm planning to make it so that you can set an option in `/etc/fscrypt.conf` to use the new keyring, if you don't mind that `fscrypt unlock` will start requiring root.  For new encrypted directories, there will be a new encryption policy version that will always use the new keyring and won't require root.

**Comment by Redsandro** [comment]
@ebiggers thank you for the update. Do I understand correctly that (after these updates) we can simply update our encrypted home, even if it's on the same device (that we previously used `fscrypt setup` on)?

E.g.:

```bash
mv /home/user /home/user-bup
mkdir /home/user
fscrypt encrypt /home/user
rsync -vaH /home/user-bup/ /home/user/
rm -rf /home/user-bup
```

Or will the renaming cause problems? (This might be a stupid question but please bear in mind that I don't understand what's going on inside `/home/.fscrypt/`.)

**Comment by ebiggers** [comment]
Yes, assuming all the kernel and userspace patches go in as planned, the best way to fix the problem will be to set `"policy_version": "2"` in the options section of  `/etc/fscrypt.conf`, then re-create your encrypted home directory as you showed.  (But please don't actually do this yet unless you're really adventurous; I'm still actively working on the userspace patches!)

Eventually we may make `fscrypt encrypt` create a v2 policy automatically if the kernel supports it, without needing to opt-in.  But I think we'll make it opt-in initially.

I'm also planning to make it possible to opt-in to the filesystem keyring for v1 encrypted directories by setting `"use_fs_keyring_for_v1_policies": true` in `/etc/fscrypt.conf`.  But that will have some disadvantages, like making `fscrypt unlock` start requiring root on those directories.

**Comment by zabbal** [comment]
Could you clarify what "requiring root" means in this context? Like an actual logged-in root user or just a "sudo fscrypt..." or particular set of capabilities?

Just out of curiosity though, I do realize that it's on-going development and things are still in flux.

**Comment by ebiggers** [comment]
I mean you'd need `sudo fscrypt unlock` to unlock the directories --- or have the PAM module unlock them using your login passphrase.  Regular users will still be able to use the directories as long as they were unlocked by `sudo` (or by something else running as root) first.

That's only for the existing "v1" encrypted directories.  Directories using the new v2 encryption policies will be unlockable as regular users, and they always use the filesystem-level keyring so they always avoid the key visibility issue with sshd/docker/NetworkManager/etc.

**Comment by ebiggers** [comment]
The solution to this problem was merged in #148, so I'm closing this.

However, due to the prerequisite of kernel v5.4 or later, the fix is currently "opt-in" via a setting in `/etc/fscrypt.conf`.  See the [new Troubleshooting section](https://github.com/google/fscrypt#cant-log-in-with-ssh-even-when-users-encrypted-home-directory-is-unlocked) for how to enable it.

https://github.com/google/fscrypt/issues/182 tracks making new installations of `fscrypt` use v2 encryption policies by default when kernel support is detected.

**Comment by Redsandro** [comment]
Thank you @ebiggers :+1: 

I will probably wait for Linux Mint 20 (Ubuntu 20.04 plus one month or so) before trying this out, assuming it comes with Linux 5.4. Until then, I'm relying on a method that is only secure as long as no one had physical access to the device.
